Incorrect teardown after mocking return value of classmethod implemented in base class

Following tests should pass
```
from flexmock import flexmock


class A:

    @classmethod
    def foo(cls, value: str):
        return value


class B(A):
    pass


def test_foo_return():
    flexmock(B).should_receive('foo').and_return('a')
    B.foo('')

    flexmock(B).should_receive('foo').and_return('b')
    B.foo('')


def test_foo_args():
    flexmock(B).should_receive('foo').with_args('a')
    B.foo('a')
```
but the second test fails with:
```
E       flexmock.MethodSignatureError: foo()
E       Did not match expectation foo(value="a")
```
For this issue to happen following conditions must be met:
- `classmethod` is implemented in base class and derived class is mocked
- before failing test there is a test that mocks return value of the function twice. If the return value is mocked only once then everything works
- second test must call the method on class not on instance
If everything is done in single test then it works as expected. The failure also does not happen if the mocking of return value is split into separate tests.

This issue started happening with `0.10.5` and is present since then.

--- a/src/flexmock/_api.py
+++ b/src/flexmock/_api.py
@@ -531,7 +531,10 @@ def flexmock_teardown() -> None:
     classes = []
     for mock_object, expectations in FlexmockContainer.flexmock_objects.items():
         saved[mock_object] = expectations[:]
-        for expectation in expectations:
+        # Reset the expectations in reversed order to ensure everything is restored to
+        # its original form even if setting consequent expectations on the same object
+        # operates on top of previous expectation and not on original object.
+        for expectation in reversed(expectations):
             expectation._reset()
         for expectation in expectations:
             # Remove method type attributes set by flexmock. This needs to be done after

--- a/tests/features/derived.py
+++ b/tests/features/derived.py
@@ -2,6 +2,7 @@
 
 # pylint: disable=missing-docstring,no-member
 from flexmock import exceptions, flexmock
+from flexmock._api import flexmock_teardown
 from tests.some_module import DerivedClass, SomeClass
 from tests.utils import assert_raises
 
@@ -128,3 +129,19 @@ class DerivedTestCase:
         ).once()  # should be twice
         assert DerivedClass().static_method_with_args(4) == 4
         assert DerivedClass.static_method_with_args(4) == 4
+
+    def test_multi_mock_class_method_on_derived_class_is_properly_torn_down(self):
+        flexmock(DerivedClass).should_receive("class_method_with_args").with_args(1).once()
+        DerivedClass.class_method_with_args(1)
+
+        flexmock(DerivedClass).should_receive("class_method_with_args").with_args(2).once()
+        DerivedClass.class_method_with_args(2)
+
+        flexmock(DerivedClass).should_receive("class_method_with_args").with_args(3).once()
+        DerivedClass.class_method_with_args(3)
+
+        # Run teardown manually to allow this test case to be contained in single test
+        flexmock_teardown()
+
+        flexmock(DerivedClass).should_receive("class_method_with_args").with_args(4).once()
+        DerivedClass.class_method_with_args(4)

